src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs

```diff
@@ -0,0 +1,389 @@
+            var messageContent = _messageStream.ToArray();
+            if (messageContent.Length > 0)
+            {
+                _requestMessage.Content = new ByteArrayContent(messageContent);
+
+                // Apply cached "Content" header values
+                foreach (var contentHeader in _contentHeaderValueCache)
+                {
+                    _requestMessage.Content.Headers.Add(contentHeader.Key, contentHeader.Value);
+                }
+            }
+
+            _requestMessage.Method = new HttpMethod(this.ActualMethod);
+
+            var send = _client.SendAsync(_requestMessage);
+            return send;
+        }
+
+        private static IDictionary<string, string> HttpHeadersToStringDictionary(HttpHeaders headers)
+        {
+            return headers.ToDictionary((h1) => h1.Key, (h2) => string.Join(",", h2.Value));
+        }
+
+        private static HttpWebResponseMessage ConvertHttpClientResponse(HttpResponseMessage response)
+        {
+            var allHeaders = HttpHeadersToStringDictionary(response.Headers).Concat(HttpHeadersToStringDictionary(response.Content.Headers));
+            return new HttpWebResponseMessage(
+                allHeaders.ToDictionary((h1) => h1.Key, (h2) => h2.Value),
+                (int)response.StatusCode,
+                () =>
+                {
+                    var task = response.Content.ReadAsStreamAsync();
+                    task.Wait();
+                    return task.Result;
+                });
+        }
+
+        private static T UnwrapAggregateException<T>(Func<T> action)
+        {
+            try
+            {
+                return action();
+            }
+            catch (AggregateException aggregateException)
+            {
+                if (aggregateException.InnerExceptions.Count == 1)
+                {
+                    throw ConvertToDataServiceWebException(aggregateException.InnerExceptions.Single() as WebException);
+                }
+
+                throw;
+            }
+        }
+
+        private static DataServiceTransportException ConvertToDataServiceWebException(WebException webException)
+        {
+            HttpWebResponseMessage errorResponseMessage = null;
+            if (webException.Response != null)
+            {
+                var httpResponse = (HttpWebResponse)webException.Response;
+                errorResponseMessage = new HttpWebResponseMessage(httpResponse);
+            }
+
+            return new DataServiceTransportException(errorResponseMessage, webException);
+        }
+
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Perform the actual cleanup work.
+        /// </summary>
+        /// <param name="disposing">If 'true' this method is called from user code; if 'false' it is called by the runtime.</param>
+        protected virtual void Dispose(bool disposing)

```

Comment:
The param disposing is abit confusing, I think it should better describe what it means. And where is it used ?